### PR TITLE
Pin all docker images to digests

### PR DESCRIPTION
**Pull request by GMNGeoffrey**
This dramatically increases the reproducibility of Docker builds and
avoids painful debugging experiences due to inconsistent Docker
caching. Note I was able to delete the entire "Debugging" section. I
was also able to remove basically all the arguments to
`manage_images.py`.

The cost is that using the script now always uploads to GCR which takes
time and requires permissions. Given that this is required to test the
images anyway, this seems not too great a cost. Personally, I've spent
way more time dealing with weird Docker edge cases than pushing images.
People can always fall back to standard Docker commands for local
iteration.

### CODE REVIEW

File: build_tools/docker/README.md [hunk 1]
```diff
@@ -38,42 +38,28 @@ dependencies based on these image names.
 We use a helper python script to manage the Docker image deployment. It lists
 all images and their dependencies and manages their canonical registry location.
 When creating a new image, add it to this mapping. To build an image and all
-images it depends on:
+images it depends on as well as pushing them to GCR and updating all references
+to the image digest. Pushing images to GCR requires the `Storage Admin` role in
+the `iree-oss` GCP project.
```
Comment (phoenix-meadowlark) [inline build_tools/docker/README.md:43]: Add a note about needing to request this for OSS contributors?

We might also want to include a brief section on pulling and running images, since that isn't possible via the script without these permissions. Something like:


## Running Images Locally

Production versions of the images can be download and run locally as follows:

```shell
PROD_IMAGE_URL="gcr.io/iree-oss/bazel-tensorflow-swiftshader:prod"
docker pull "${PROD_IMAGE_URL?}"
docker run -it --rm --entrypoint bash "${PROD_IMAGE_URL?}"
```

**Review by phoenix-meadowlark — CHANGES_REQUESTED** [review]
Love how much this simplifies things :)

File: build_tools/docker/manage_images.py [hunk 2]
```diff
@@ -221,51 +222,67 @@ def update_references(image_url: str, digest: str, dry_run: bool = False):
...
+
+
+# TODO typing
+def get_prod_digests() -> Dict[str, str]:
+  with open(utils.PROD_DIGESTS_PATH, "r") as f:
+    images_with_digests = {}
+    for line in f:
+      url, digest = line.strip().split("@")
+      images_with_digests[url] = digest
+  return images_with_digests
```
Comment (phoenix-meadowlark) [inline build_tools/docker/manage_images.py:239]: Ugh, not sure how my suggestion got deleted from the review I sent :|

```suggestion
def parse_prod_digests() -> Dict[str, str]:
  image_urls_to_prod_digests = {}
  with open(utils.PROD_DIGESTS_PATH, "r") as f:
    for line in f:
      image_url, digest = line.strip().split("@")
      image_urls_to_prod_digests[image_url] = digest
  return image_urls_to_prod_digests
```

File: build_tools/docker/manage_images.py [hunk 3]
```diff
@@ -221,51 +222,67 @@ def update_references(image_url: str, digest: str, dry_run: bool = False):
...
-    # Ensure the user has the correct authorization if they try to push to GCR.
-    utils.check_gcloud_auth(dry_run=args.dry_run)
+  # Ensure the user has the correct authorization to push to GCR.
+  utils.check_gcloud_auth(dry_run=args.dry_run)
 
   images_to_process = get_ordered_images_to_process(args.images)
   print(f'Also processing dependent images. Will process: {images_to_process}')
 
+  dependencies = get_dependencies(images_to_process)
+  print(f'Pulling image dependencies: {list(dependencies)}')
```
[the remark this answers is not in the document]
Comment (GMNGeoffrey) [reply build_tools/docker/manage_images.py:252]: Oh at some point I had dependencies as a set, but then it upset typing and I decided I didn't care.

File: build_tools/docker/manage_images.py [hunk 4]
```diff
@@ -142,6 +125,21 @@ def add_dependent_images(image: str):
...
+  def add_dependency(image: str):
+    if image not in dependencies:
+      for dependency in IMAGES_TO_DEPENDENCIES[image]:
+        add_dependency(dependency)
+      dependencies.append(image)
+
+  for image in images:
+    add_dependency(image)
+
+  return dependencies
```
[the remark this answers is not in the document]
Comment (GMNGeoffrey) [reply build_tools/docker/manage_images.py:140]: Sure. At least in Java land, it's common to not promise something stricter than what you want to guarantee, but this is a private function so who cares.

File: build_tools/docker/manage_images.py [hunk 5]
```diff
@@ -189,10 +187,13 @@ def get_repo_digest(tagged_image_url: str) -> str:
...
+  for line in fileinput.input(files=['WORKSPACE'], inplace=True):
     if line.strip().startswith('digest ='):
       digest_updates += 1
-      print(re.sub(DIGEST_REGEX, digest, line), end='')
+      if dry_run:
+        print(line, end='')
+      else:
+        print(re.sub(DIGEST_REGEX, digest, line), end='')
     else:
       print(line, end='')
```
[the remark this answers is not in the document]
Comment (GMNGeoffrey) [reply build_tools/docker/manage_images.py:198]: Yeah maybe. Can that be a followup?

[the remark this answers is not in the document]
Comment (phoenix-meadowlark) [reply build_tools/docker/manage_images.py:208]: SGTM

File: build_tools/docker/README.md (hunk 1 above)
Comment (GMNGeoffrey) [reply build_tools/docker/README.md]: Done

**Review by phoenix-meadowlark — APPROVED** [review]
